cpp/src/copying/segmented_shift.cu

```diff
@@ -0,0 +1,281 @@
+                                                    size_type offset,
+                                                    device_span<size_type const> segment_offsets,
+                                                    size_type column_size,
+                                                    rmm::cuda_stream_view stream,
+                                                    rmm::mr::device_memory_resource* mr)
+{
+  if (offset > 0) {
+    auto filter = [segment_offsets, offset] __device__(auto const& i) {
+      auto segment_bound_idx =
+        thrust::upper_bound(thrust::seq, segment_offsets.begin(), segment_offsets.end(), i) - 1;
+      return not(*segment_bound_idx <= i and i < *segment_bound_idx + offset);
+    };
+    return strings::detail::copy_if_else(input_pair_iterator,
+                                         input_pair_iterator + column_size,
+                                         fill_pair_iterator,
+                                         filter,
+                                         stream,
+                                         mr);
+  } else {
+    auto filter = [segment_offsets, offset] __device__(auto const& i) {
+      auto segment_bound_idx =
+        thrust::upper_bound(thrust::seq, segment_offsets.begin(), segment_offsets.end(), i);
+      return not(*segment_bound_idx + offset <= i and i < *segment_bound_idx);
+    };
+    return strings::detail::copy_if_else(input_pair_iterator,
+                                         input_pair_iterator + column_size,
+                                         fill_pair_iterator,
+                                         filter,
+                                         stream,
+                                         mr);
+  }
+}
+
+template <typename T, typename Enable = void>
+struct segmented_shift_functor {
+  template <typename... Args>
+  std::unique_ptr<column> operator()(Args&&...)
+  {
+    CUDF_FAIL("Unsupported type for segmented_shift.");
+  }
+};
+
+/**
+ * @brief Segmented shift specialization for representation layout compatible types.
+ */
+template <typename T>
+struct segmented_shift_functor<T, std::enable_if_t<is_rep_layout_compatible<T>()>> {
+  std::unique_ptr<column> operator()(column_view const& segmented_values,
+                                     device_span<size_type const> segment_offsets,
+                                     size_type offset,
+                                     scalar const& fill_value,
+                                     rmm::cuda_stream_view stream,
+                                     rmm::mr::device_memory_resource* mr)
+  {
+    auto values_device_view = column_device_view::create(segmented_values, stream);
+    auto fill_pair_iterator = make_pair_iterator<T>(fill_value);
+    bool nullable           = not fill_value.is_valid() or segmented_values.nullable();
+
+    if (segmented_values.has_nulls()) {
+      auto input_pair_iterator = make_pair_iterator<T, true>(*values_device_view) - offset;
```

Comment:
Can this use the new optional_iterator instead of pair_iterator?

Reply:
So it won't be a simple drop in replacement. Currently `copy_if_else` expects its input to be a pair_iterator:
https://github.com/rapidsai/cudf/blob/791beb2989906bd48ea743a294ab56e4fb0bf6de/cpp/include/cudf/detail/copy_if_else.cuh#L144-L147